<?xml version="1.0" encoding="UTF-8"?>
<!-- SPDX-License-Identifier: BSD-3-Clause -->
<!-- SPDX-FileCopyrightText: Czech Technical University in Prague -->

<!-- Run magnetometer compass to provide azimuth. -->

<launch>
    <arg name="frame" default="base_link" description="The frame in which the azimuth should be computed." />
    <arg name="unbias" default="true" description="When true, magnetometer bias observer and remover will be launched." />
    <arg name="low_pass_ratio" default="0.93" description="Low pass smoothing ratio. 0 = raw measurements, 1 = no updates." />
    <arg name="initial_variance" default="1e-1" description="Compass uses static covariance for now." />
    <arg name="visualize" default="true" description="Whether to publish a visualization of the azimuth" />
    <arg name="magnetic_declination" default="-0.068" description="Set magnetic declination to zero in simulation." />
    <!-- <arg name="initial_lat" default="50.25169" description="for bag file without gps fix." /> -->
    <!-- <arg name="initial_lon" default="13.12277" description="for bag file without gps fix." /> -->

    <include file="$(find-pkg-share magnetometer_pipeline)/launch/magnetometer_proc_launch_bag.xml" if="$(var unbias)">
    </include>

    <node pkg="magnetometer_compass" exec="magnetometer_compass_nodelet" name="compass" respawn="true">
        <param name="frame" value="$(var frame)" />
        <param name="low_pass_ratio" value="$(var low_pass_ratio)" />
        <param name="magnetic_declination" value="$(var magnetic_declination)" />
        <param name="initial_variance" value="$(var initial_variance)" />
        <param name="subscribe_mag_unbiased" value="$(var unbias)" />
        <param name="publish_mag_azimuth_enu_rad" value="true" />
        <param name="publish_mag_azimuth_enu_imu" value="true" />
    </node>

    <node pkg="magnetometer_compass" exec="visualize_azimuth_nodelet" name="visualize_azimuth" if="$(var visualize)">
        <remap from="visualize_azimuth/azimuth" to="compass/mag/enu/rad" />
        <remap from="gps/fix" to="navsatfix" />
        <param name="magnetic_declination" value="$(var magnetic_declination)" />
        <param name="max_rate" value="10." />
        <param name="input_orientation" value="enu" />
        <param name="input_reference" value="magnetic" />
        <!-- <param name="initial_lat" value="$(var initial_lat)" />
        <param name="initial_lon" value="$(var initial_lon)" /> -->
    </node>
</launch>